### ISSUE TITLE
Failed to build Tensorflow Lite for Win32 using CMake

### ISSUE THREAD

**Issue by tcervi** [tensorflow/tensorflow#47166]
**System information**
- OS Platform and Distribution: Windows - 10.0.19041 - AMD64
- TensorFlow installed from: tag v2.4.1 (85c8b2a) or master (bf157579cf0)
- TensorFlow version: 2.4.1
- CMake version: 3.17.2
- GCC/Compiler version: MSVC 19.27.29111.0
- CUDA/cuDNN version: Not used
- GPU model and memory: Not used

This is probably not relevant since I'm building with CMake, but follow my Bazel environment information:
- Python version: 3.8.3 (x64 and x86)
- Installed using virtualenv: tried with both Python x86 venv and with x64 venv
- Bazel version: 3.1.0

**My problem:**
I'm currently tying to build Tensorflow Lite library to be used by my library that should support both Linux and Windows, x86 and x64. Since Bazel does not builds for x86 out of the box I was following the [documentation](https://www.tensorflow.org/lite/guide/build_cmake) to build TF Lite using CMake, including its source with _add_subdirectory()_ .
It works great for for x64 builds (-A x64) but when I change it to windows x86 (-A Win32) it always fails building _gemmlowp_.
I also tried to build via cmake on command line, using the tensorflow/lite/CMakeLists.txt to build the static library, but get the same error.
**I wonder if it was even supposed to be supported** and if so, if it is a issue to be handled by tensorflow or [gemmlowp](https://github.com/google/gemmlowp). I can also see some pthread releated errors on CMakeError.log

**Build steps:**

- λ cmake -A Win32 -DCMAKE_BUILD_TYPE=Release -S ..\tensorflow\lite\ -B .
- λ cmake --build . --config Release
...
  ...
  expand_dims.cc
  fake_quant.cc
  Generating Code...
**C:\workspace\tensorflow\build-tflite\gemmlowp\internal\output.h(176): fatal error C1001: Internal compiler error.** [C :\workspace\tensorflow\build-tflite\tensorflow-lite.vcxproj]
  (compiler file 'd:\agent\_work\7\s\src\vctools\Compiler\Utc\src\p2\main.c', line 195)
   To work around this problem, try simplifying or changing the program near the locations listed above.
  If possible please provide a repro here: https://developercommunity.visualstudio.com
  Please choose the Technical Support command on the Visual C++
   Help menu, or open the Technical Support help file for more information

**Other logs**

**CMakeError.log**

Determining if the include file pthread.h exists failed with the following output:
Change Dir: C:/workspace/tensorflow/build-tflite/CMakeFiles/CMakeTmp

...

Copyright (C) Microsoft Corporation. All rights reserved.

  Microsoft (R) C/C++ Optimizing Compiler Version 19.27.29111 for x86

  CheckIncludeFile.c

  Copyright (C) Microsoft Corporation.  All rights reserved.

  cl /c /I"C:\workspace\vcpkg\scripts\buildsystems\msbuild\..\..\..\installed\x86-windows\include" /Zi /W1 /WX- /diagnostics:column /Od /Ob0 /Oy- /D WIN32 /D _WINDOWS /D "CMAKE_INTDIR=\"Debug\"" /D _MBCS /Gm- /RTC1 /MDd /GS /fp:precise /Zc:wchar_t /Zc:forScope /Zc:inline /Fo"cmTC_d958e.dir\Debug\\" /Fd"cmTC_d958e.dir\Debug\vc142.pdb" /Gd /TC /analyze- /errorReport:queue "C:\workspace\tensorflow\build-tflite\CMakeFiles\CMakeTmp\CheckIncludeFile.c"

C:\workspace\tensorflow\build-tflite\CMakeFiles\CMakeTmp\CheckIncludeFile.c(1,10): **fatal error C1083: Cannot open include file: 'pthread.h': No such file or directory** [C:\workspace\tensorflow\build-tflite\CMakeFiles\CMakeTmp\cmTC_d958e.vcxproj]

Performing C++ SOURCE FILE Test EIGEN_COMPILER_SUPPORT_CPP11 failed with the following output:
Change Dir: C:/workspace/tensorflow/build-tflite/CMakeFiles/CMakeTmp

...

  Microsoft (R) C/C++ Optimizing Compiler Version 19.27.29111 for x86
  src.cxx
  Copyright (C) Microsoft Corporation.  All rights reserved.

  cl /c /I"C:\workspace\vcpkg\scripts\buildsystems\msbuild\..\..\..\installed\x86-windows\include" /Zi /W1 /WX- /diagnostics:column /Od /Ob0 /Oy- /D WIN32 /D _WINDOWS /D EIGEN_COMPILER_SUPPORT_CPP11 /D "CMAKE_INTDIR=\"Debug\"" /D _MBCS /Gm- /EHsc /RTC1 /MDd /GS /fp:precise /Zc:wchar_t /Zc:forScope /Zc:inline /GR /Fo"cmTC_94317.dir\Debug\\" /Fd"cmTC_94317.dir\Debug\vc142.pdb" /Gd /TP /analyze- /errorReport:queue  -std=c++11 "C:\workspace\tensorflow\build-tflite\CMakeFiles\CMakeTmp\src.cxx"

cl : **command line warning D9002: ignoring unknown option '-std=c++11'** [C:\workspace\tensorflow\build-tflite\CMakeFiles\CMakeTmp\cmTC_94317.vcxproj]

Source file was:
int main(int argc, char* argv[]) { return (int)__builtin_expect(0, 0); }
**Checking whether the ASM compiler is GNU using "--version" did not match "(GNU assembler)|(GCC)|(Free Software Foundation)":**
Microsoft (R) C/C++ Optimizing Compiler Version 19.27.29111 for x86
Copyright (C) Microsoft Corporation.  All rights reserved.

Thanks in advance, cheers!

**Comment by terryheo** [comment]
Hi, I tried this and it works well on my side.

I used CMake 3.20.0-rc2 and MSVC 19.27.29111.0.
Also I installed some necessary packages for TF build. https://www.tensorflow.org/install/source_windows

```
C:\tensorflow\tfbuild>cmake -A Win32 ..\tensorflow\lite -DTFLITE_ENABLE_XNNPACK=OFF
-- Building for: Visual Studio 16 2019
-- Setting build type to Release, for debug builds use'-DCMAKE_BUILD_TYPE=Debug'.
-- The C compiler identification is MSVC 19.27.29111.0
-- The CXX compiler identification is MSVC 19.27.29111.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: C:/Program Files (x86)/Microsoft Visual Studio/2019/Community/VC/Tools/MSVC/14.27.29110/bin/Hostx64/x86/cl.exe - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: C:/Program Files (x86)/Microsoft Visual Studio/2019/Community/VC/Tools/MSVC/14.27.29110/bin/Hostx64/x86/cl.exe - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Looking for pthread.h
-- Looking for pthread.h - not found
-- Found Threads: TRUE
```

And I verified it's a 32bit binary.
```
C:\tensorflow\tfbuild>debug\benchmark_model
STARTING!
Duplicate flags: num_threads
Please specify the name of your TF Lite input file with --graph
Benchmarking failed.

C:\tensorflow\tfbuild>file debug\benchmark_model.exe
debug\benchmark_model.exe: PE32 executable (console) Intel 80386, for MS Windows
```

**Comment by tcervi** [comment]
Thanks for your feedback @terryheo,
from which branch/tag did you run this build?

Will give it a new try next Monday and check if I missed any Windows dependency.

Cheers

**Comment by terryheo** [comment]
I used master branch.

**Comment by tcervi** [comment]
Indeed it works from master branch @terryheo
It only generates tensorflow-lite.lib though, for both x64 and x86 (Win32) 😢 
I probably was not clear before, but my main issue is that I need to generate a Dynamic library for Win32 systems.

I took a look at tensorflow/lite/CMakeLists;txt and forced to generate a shared line adding SHARED to line 371:
`add_library(tensorflow-lite SHARED`
I used this change to build a x64 dll and compare to the one built using Bazel.
It generated a dll but much smaller than the dll generate when using Bazel to build `//tensorflow/lite:tensorflowlite.dll`
Using Bazel the output dll has ~14Mb and ~30k exported symbols. The CMake output dll has ~3Mb and 33 exported symbols.
Apparently the CMake output does not contain the library dependencies. In fact when I tried to dynamically use it windows complained about Flatbuffers symbols that could not be located in the dynamic link library, and am using it the same way I do with Bazel output dll. 

That's why I assume the dependencies are note being handled properly and will dig a bit deeper into this CMake build to compare the differences when building via Bazel, but would also appreciate any help or thoughts from your side. Again, Bazel would not help me to build it for Win32 systems 😢 

I checked both dll dependencies using Dumpbin, the results follow:

- CMake output dll:
`>dumpbin /dependents tensorflow-lite.dll | findstr dll
Dump of file tensorflow-lite.dll
KERNEL32.dll`
- Bazel output dll:
`>dumpbin /dependents tensorflowlite.dll | findstr dll
Dump of file tensorflow-lite.dll
MSVCP140.dll
KERNEL32.dll
VCRUNTIME140.dll
VCRUNTIME140_1.dll
api-ms-win-crt-runtime-l1-1-0.dll
api-ms-win-crt-math-l1-1-0.dll
api-ms-win-crt-heap-l1-1-0.dll
api-ms-win-crt-utility-l1-1-0.dll
api-ms-win-crt-stdio-l1-1-0.dll
api-ms-win-crt-string-l1-1-0.dll
api-ms-win-crt-convert-l1-1-0.dll
api-ms-win-crt-filesystem-l1-1-0.dll
api-ms-win-crt-environment-l1-1-0.dll
api-ms-win-crt-locale-l1-1-0.dll`

I used DependencyWalker to get this screenshot:
![image](https://user-images.githubusercontent.com/19482569/111505405-59808000-8727-11eb-9ece-13bc65480201.png)

**Comment by terryheo** [comment]
If you want to use a shared library, how about using C API?
https://www.tensorflow.org/lite/guide/build_cmake#build_tensorflow_lite_c_library

Even though you have a C++ standalone shared library, you might have trouble to collect all the necessary header files.
(If you're not having a CMake project)
In this case, using C API is much easier.

**Comment by tcervi** [comment]
**About the C++ standalone shared library:** 
I reckon I got all the header files sorted since was able to build and run x64 version. My issue was building the shared library for x86.
**About the C API:** 
It's indeed a good suggestion, for now my use case is simple (load model, load input tensor, run inference and read output tensor), it should handle it. 

I did compile the C API using CMake and MSVC for x64 and Win32 (Debug and Release) but there is something wrong with the Win32 release build when it tries to build gemmlowp related code. I reckon it is the same issue I faced in the beginning when I raised this issue here. I just can not understand why it works for Win32 Debug builds, need to investigate further 😕 

**Results** - TFLite C API (*.dll) Build: 
- MSVC x64 Release: ✅ 
- MSVC x64 Debug: ✅ 
- MSVC x86 Release: ❌ 
- MSVC x86 Debug: ✅ 

**Output Error for Win32 Release build:**
- λ cmake -A Win32 ../tensorflow/lite/c 
(with or without -DTFLITE_ENABLE_XNNPACK=OFF)
- λ cmake --build . --config Release
...
  expand_dims.cc
  fake_quant.cc
  Generating Code...
C:\workspace\tensorflow\build-tflite\gemmlowp\internal\output.h(176): fatal error C1001: Internal compiler error. [C:\workspa
ce\tensorflow\build-tflite\tensorflow-lite\tensorflow-lite.vcxproj]
  (compiler file 'd:\agent\_work\7\s\src\vctools\Compiler\Utc\src\p2\main.c', line 195)
   To work around this problem, try simplifying or changing the program near the locations listed above.
  If possible please provide a repro here: https://developercommunity.visualstudio.com
  Please choose the Technical Support command on the Visual C++
   Help menu, or open the Technical Support help file for more information

The line 176 is the end of a gemmlowp enum, not sure what is causing this build error. I'm specially confused because the build works when it is a Debug build 😄 
My problem with Debug DLL is that the inference time ridiculously increased 😢 


**To sum it up:**
Thanks @terryheo for the C API tip, it may make my life easier indeed.
The CMake build for C API with MSVC targetin Win32 still have an issue though.
Not sure if the build issue should be addressed to TF team or to Gemmlowp team. Also, am not sure if it would even be considered to be fixed. The fix would help people trying to use TFLite on Win32 devices but probably that's a vast minority of TF users. Unfortunately, since my project is multi-platform and multi-arch, I'm one of those 😄

**Comment by tcervi** [comment]
Hi @ruiyq, thanks for your comment, indeed your configuration works!
When I searched for that  MSVC "C1001: Internal compiler error." found [some information](https://docs.microsoft.com/en-us/cpp/error-messages/compiler-errors-1/fatal-error-c1001?view=msvc-160) about this error being usually linked to optimisation failure. I was planning to maybe generate a build repro and upload it for MSVC team, but was not investing that much time on this specific build and the simple change from "/O2" to "/O1", as you suggested, solved my issue. In my case the speed decrease is negligible. 
Thanks again for sharing your experience!

@terryheo since I was not using Visual Studio Code, instead was building from command line using cmake + msvc build tools (windows container), there was no gui for me to configure the build optimisation flags. I solved it by modifying the CMakeLists.txt: checking for MSVC-x86 builds and updating the optimisation flags accordingly. Do you think it is a valid PR?
If in the future the MSVC team fix something on their compiler and it becomes able to build tflite-c with "/O2", the code can be removed. Until there, anyone using MSVC to build for Win32 targets will not suffer with **THIS** build failure.
I guess the main goal of Tensorflow Lite is the portability right?

Cheers

**Comment by Kartik14** [comment]
Hi @tcervi I am also trying to build a shared tflite c library for x86 and ARM64. Using the fix you sent as a pull request, the build process gets completed successfully, But in the Release directory, only the .dll file generated(no .lib file). And the size of the .dll file is also too small (180 KB). 

I see that in one of the [comment ](https://github.com/tensorflow/tensorflow/issues/47166#issuecomment-801250436) above, you also encounter a similar issue for C++ builds. Are you able to get the ".lib" file for C library build?

Thanks

**Comment by terryheo** [comment]
@Kartik14 did you follow 
https://www.tensorflow.org/lite/guide/build_cmake#build_tensorflow_lite_c_library ?

**Comment by Kartik14** [comment]
Hi @terryheo Thanks for the quick response. I used slightly different commands, similar to @tcervi 

1) First I changed the **CMakeLists.txt** file in **tensorflow\lite\c** according to the pull request [here](https://github.com/tensorflow/tensorflow/pull/49171)

2) Then I used the following two commands:

`cmake -A Win32 -DCMAKE_BUILD_TYPE=Release -S ..\tensorflow\lite\c -B .`
`cmake --build . --config Release`

It generated a **tensorflowlite_c.dl**l file in the **Release** folder. But isn't it also supposed to create a **.lib** file (something like tensorflowlite_c.dll.if.lib) for use in some other project? I apologize if I am asking a wrong question since it is my first time trying to build a shared C library. 

Thanks

**Comment by terryheo** [comment]
@Kartik14
As showed in the https://www.tensorflow.org/lite/guide/build_cmake#build_tensorflow_lite_c_library 
tensorflowlite_c.dll is the target binary you will use.
It doesn't generate lib file since it's difficult to use for 3rd party library dependency (without using CMake project).

Do you have any issue on using the dll file?

**Comment by Kartik14** [comment]
@terryheo 

I am using Visual Studio to build my other project (where I wish to use tflite). I am using steps described [here](https://docs.microsoft.com/en-us/cpp/build/walkthrough-creating-and-using-a-dynamic-link-library-cpp?view=msvc-160) to import the dll file in my project. But here they [add the lib file](https://docs.microsoft.com/en-us/cpp/build/walkthrough-creating-and-using-a-dynamic-link-library-cpp?view=msvc-160#to-add-the-dll-import-library-to-your-project) as additional dependencies.

I tried to find if there was a way to use just the dll file, but [here](https://stackoverflow.com/questions/38336876/dont-know-how-to-load-and-use-dll-file-in-c) the user mentions:

>Generally, the library will consist of a .lib file (for linking to your .exe), a .dll file (for dynamically linked code) and an include >folder with lots of .h files. You cannot use a .dll file on its own, you must have the headers and lib files.

Can you please guide to some resource I can check out for importing the dll file.

Thanks

**Comment by tcervi** [comment]
Hi @Kartik14 

I was not using Visual Studio, only building my library windows version with a combination of MSVC and CMake. And yes I used the built DLL (x86 release) along with my library and for my use it worked fine (model inferences).
Using the PR code I was able to build both .lib and .dll.

Previously, when I was building from commit [bf15757](https://github.com/tensorflow/tensorflow/commit/bf157579cf016bcd00e4d01956a0807a2ff6f608) + my PR code, MSVC was finishing the build with both .lib and .dll on the same dir. Recently (master branch up to date), not sure if it was a TF code update or MSVC version update, it started to release both files in separate dirs.
I can't remember which sub directory now, but it should definitely be inside your build output dir. Anyways you can give a try on that commit SHA I linked and check the output as well.

Last question: did you run the configuration script before building it? 
If not, please check this link: https://www.tensorflow.org/install/source_windows

**Comment by Kartik14** [comment]
Hi @tcervi, I was indeed able to build both x64 and x86 (dll and lib file) using the commit SHA you linked. Thanks a lot!

**Comment by tcervi** [comment]
Cool stuff @Kartik14,
were you able to build both from main branch too? 
If I'm not wrong the build should generate a "tensorflow-lite" directory  with "Release" and the .lib inside.

**Comment by tcervi** [comment]
Solved by: https://github.com/tensorflow/tensorflow/pull/49171

**Comment by aaghiijnnuz** [comment]
I can compile with the command below

**Comment by aaghiijnnuz** [comment]
I can compile

**Comment by aaghiijnnuz** [comment]
I can compile with te command below

For x64
D:\Downloads\tensorflow-r2.6\buildGui64>cmake ..\tensorflow\lite\c -DTFLITE_ENABLE_XNNPACK=OFF -DCMAKE_WINDOWS_EXPORT_ALL_SYMBOLS=True

D:\Downloads\tensorflow-r2.6\buildGui>cmake -A Win32 ..\tensorflow\lite\c -DCMAKE_WINDOWS_EXPORT_ALL_SYMBOLS=True -DTFLITE_ENABLE_XNNPACK=OFF

**Comment by Yamilemon** [comment]
Is there anyone has tensorflowlite_c.dll and tensorflowlite_c.dll.if.lib which can run under x86? I need it. I found x86_64 here: [](https://github.com/ValYouW/tflite-dist), but this cannot be used in X86 project in visual studio.

**Comment by Yamilemon** [comment]
Is there anyone has tensorflowlite_c.dll and tensorflowlite_c.dll.if.lib which can run under x86? I need it. I found x86_64 here: https://github.com/ValYouW/tflite-dist, but this cannot be used in X86 project in visual studio.

**Comment by apratimankit** [comment]
showing this after building x86 lib and dll 
LNK2001	unresolved external symbol __imp__TfLiteInterpreterAllocateTensors
